### PR TITLE
feat: Added LRU Cache Data Structure

### PR DESCRIPTION
**Pull request by proRamLOGO**
#### Description 
Code for LRU Cache Data Structure has been added to the data structures folder. LRU Cache has been implemented using a circular doubly linked list.
#### Checklist

- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1356"></a>

### CODE REVIEW

File: data_structures/LRUCache.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
+/**
+ * @file
+ * @brief Implementation of LRU Cache Data Structure
+ * @details
+ * This is a multi line
+ * description containing links, references,
+ * math equations, etc
```
Comment (Panquesito7) [inline data_structures/LRUCache.cpp:7]:
```suggestion
```

File: data_structures/LRUCache.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+ * A Least Recently Used (LRU) Cache organizes items in order of use, allowing you 
+ * to quickly identify which item hasn't been used for the longest amount of time.
+ * Implemented using Static Doubly Circular Linked List
+ * @author [Shubh Bansal](https://github.com/proRamLOGO)
+ * @see
```
Comment (Panquesito7) [inline data_structures/LRUCache.cpp:12]:
```suggestion
```

File: data_structures/LRUCache.cpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+ * A Least Recently Used (LRU) Cache organizes items in order of use, allowing you 
+ * to quickly identify which item hasn't been used for the longest amount of time.
+ * Implemented using Static Doubly Circular Linked List
+ * @author [Shubh Bansal](https://github.com/proRamLOGO)
+ * @see 
+ */
+
+#include <iostream>
+#include <unordered_map> // For hashmap
+#include <cassert>
```
Comment (Panquesito7) [inline data_structures/LRUCache.cpp:17]:
```suggestion
#include <iostream>  /// for io operations
#include <unordered_map> // For hashmap
#include <cassert>  /// for assert
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/1283307334) warnings.

**Review by proRamLOGO — COMMENTED** [review]
If key requested is not present in LRU Cache function get will return -1.

**Comment by proRamLOGO** [comment]
It is workin fine now. Moderator pls check

**Comment by proRamLOGO** [comment]
@Panquesito7 my code is working fine on ubuntu and macOS but not on windows-latest. Pls tell what should I do now ?

File: data_structures/LRUCache.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+ * @details
+ * A Least Recently Used (LRU) Cache organizes items in order of use, allowing you 
+ * to quickly identify which item hasn't been used for the longest amount of time.
+ * Implemented using Static Doubly Circular Linked List
+ * @author [Shubh Bansal](https://github.com/proRamLOGO)
+ */
+
+#include <iostream>  /// for io operations
+#include <unordered_map> // For hashmap
+#include <cassert>  /// for assert
```
Comment (Panquesito7) [inline data_structures/LRUCache.cpp:13]:
```suggestion
#include <iostream>      /// for io operations
#include <unordered_map> /// for hashmap
#include <cassert>       /// for assert
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/proRamLOGO/C-Plus-Plus/actions

**Comment by proRamLOGO** [comment]
> Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/proRamLOGO/C-Plus-Plus/actions

Done ENABLED

**Comment by Panquesito7** [comment]
Good work. Now, to trigger the GitHub Actions, make a small commit.
Next, we shall see a commit made by GitHub Actions updating `DIRECTORY.md`. 🙂

File: data_structures/lrucache.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
+/**
+ * @file
+ * @brief Implementation of LRU Cache Data Structure
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:3]:
```suggestion
 * @brief Implementation of [LRU Cache Data](https://www.geeksforgeeks.org/lru-cache-implementation/) Structure
```

File: data_structures/lrucache.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+ * Implemented using Static Doubly Circular Linked List
+ * @author [Shubh Bansal](https://github.com/proRamLOGO)
+ */
+
+#include <iostream>       /// for io operations
+#include <unordered_map>  /// for hashmap
+#include <cassert>        /// for assert
+
+/**
+ * Class node
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:16]:
```suggestion
```

File: data_structures/lrucache.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+ * Class node to build up and implement doubly circular linked lists
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:17]:
```suggestion
 * @brief Class node to build up and implement doubly circular linked lists
```

File: data_structures/lrucache.cpp [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ * to quickly identify which item hasn't been used for the longest amount of time.
+ * Implemented using Static Doubly Circular Linked List
+ * @author [Shubh Bansal](https://github.com/proRamLOGO)
+ */
+
+#include <iostream>       /// for io operations
+#include <unordered_map>  /// for hashmap
+#include <cassert>        /// for assert
+
+/**
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:15]:
```suggestion
/**
 * @namespace data_structures
 * @brief Algorithms with data structures
 */
namespace data_structures {
/**
 * @namespace lru_cache
 * @brief Functions for [Least Recently Used Cache](https://www.geeksforgeeks.org/lru-cache-implementation/) implementation
 */
namespace lru_cache {
/**
```

File: data_structures/lrucache.cpp [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            }
+            this->root->key = key ;
+            this->root->val = value ;
+            addr[key] = this->root ;
+            this->root = this->root->next ;
+        }
+        get(key) ;
+        addr[key]->val = value ;
+    }
+};
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:134]:
```suggestion
};
}  // namespace lru_cache
}  // namespace data_structures
```

File: data_structures/lrucache.cpp [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+   
+    LRUCache lru(2); ///< Created a new LRUCache instace of physical size 2.
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:144]:
```suggestion
    data_structures::lru_cache::LRUCache lru(2); ///< Created a new LRUCache instace of physical size 2.
```

File: data_structures/lrucache.cpp [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    int key; ///< Variable to contain value of {key}
+    int val; ///< Variable to contain value of {value} associated with {key}
+    node *next; //< Pointer to point on the node appearing next to current node.
+    node *prev ; //< Pointer to point on the node appearing previous to current node.
+
+    /**
+    * Constructor Function
+    * Accepts 2 parameters key and value for all new instaces of class node created.
+    * @param key : Key
+    * @param val : Value
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:31]: What do these parameters do? Document it here (in the code).

File: data_structures/lrucache.cpp [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        
+        // CLL of size capacity initialized.
+        
+    }    
+
+    /**
+    * get Function
+    * Accepts one parameter, the key whose value needs to be returned.
+    * If key requested is not present return -1.
+    * @param key : value for this key is to be returned.
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:84]:
```suggestion
    * @param key value for this key is to be returned.
```

File: data_structures/lrucache.cpp [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        return found->second->val ;
+    }
+    
+    /**
+    * put Function
+    * Accepts two parameters key and value. Updates the value of the key if the key exists. 
+    * Otherwise, adds the key-value pair to the cache. If the number of keys exceeds the 
+    * capacity from this operation, the function evicts the least recently used key.
+    * @param key : The key to be added or updated.
+    * @param value : Value for the respective key to be stored.
```
Comment (Panquesito7) [inline data_structures/lrucache.cpp:117]:
```suggestion
    * @param key the key to be added or updated.
    * @param value value for the respective key to be stored.
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work on updating the code. 👍
Please change the file name to `lru_cache.cpp`.

**Review by Panquesito7 — COMMENTED** [review]
Please change the file name to `lru_cache.cpp`.

File: data_structures/lru_cache.cpp [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+/**
+ * @namespace lru_cache
+ * @brief Functions for [Least Recently Used
+ * Cache](https://www.geeksforgeeks.org/lru-cache-implementation/)
+ * implementation
+ */
+namespace lru_cache {
+/**
+ * Class node
+ * Class node to build up and implement doubly circular linked lists
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:30]:
```suggestion
 * @brief Class node to build up and implement doubly circular linked lists
```

File: data_structures/lru_cache.cpp [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+namespace data_structures {
+/**
+ * @namespace lru_cache
+ * @brief Functions for [Least Recently Used
+ * Cache](https://www.geeksforgeeks.org/lru-cache-implementation/)
+ * implementation
+ */
+namespace lru_cache {
+/**
+ * Class node
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:29]:
```suggestion
```

File: data_structures/lru_cache.cpp [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+
+    /**
+     * Constructor Function
+     * Accepts 2 parameters key and value for all new instaces of class node
+     * created.
+     * @param key : Key
+     * @param val : Value
+     * @returns NOTHING
+     */
+    node(int key, int val) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:50]:
```suggestion
    node(uint64_t key, uint64_t val) {
```

File: data_structures/lru_cache.cpp [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+namespace lru_cache {
+/**
+ * Class node
+ * Class node to build up and implement doubly circular linked lists
+ * which will be used as LRU Cache DS.
+ */
+class node {
+ public:
+    int key;  ///< Variable to contain value of {key}
+    int val;  ///< Variable to contain value of {value} associated with {key}
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:36]:
```suggestion
    uint64_t key = 0;  ///< Variable to contain value of {key}
    uint64_t val = 0;  ///< Variable to contain value of {value} associated with {key}
```

File: data_structures/lru_cache.cpp [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+    std::unordered_map<int, node *> addr;
+
+    /**
+     * Constructor Function
+     * Accepts one parameter, the physical size for LRU.
+     * Creates a doubly circular linked list containg {capacity} number of
+     * nodes.
+     * @param capacity the physical size for LRU as needed for particular
+     * instance.
+     * @returns NOTHING
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:80]:
```suggestion
     * @returns nothing
```

**Comment by proRamLOGO** [comment]
> Good work on updating the code. 👍
> Please change the file name to `lru_cache.cpp`.

Done

File: data_structures/lru_cache.cpp [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+     */
+    explicit LRUCache(int capacity) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:82]:
```suggestion
    explicit LRUCache(const uint64_t& capacity) {
```

File: data_structures/lru_cache.cpp [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+
+        // CLL of size capacity initialized.
+    }
+
+    /**
+     * get Function
+     * Accepts one parameter, the key whose value needs to be returned.
+     * If key requested is not present return -1.
+     * @param key value for this key is to be returned.
+     * @returns NOTHING
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:101]: This function **does** return something.
Please document it well, make it understandable.

File: data_structures/lru_cache.cpp [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+     */
+    int get(int key) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:103]:
```suggestion
    uint64_t get(uint64_t key) {
```

File: data_structures/lru_cache.cpp [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+     * put Function
+     * Accepts two parameters key and value. Updates the value of the key if the
+     * key exists. Otherwise, adds the key-value pair to the cache. If the
+     * number of keys exceeds the capacity from this operation, the function
+     * evicts the least recently used key.
+     * @param key the key to be added or updated.
+     * @param value value for the respective key to be stored.
+     * @returns NOTHING
+     */
+    void put(int key, int value) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:138]:
```suggestion
    void put(uint64_t key, const uint64_t& value) {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work. 👍

File: data_structures/lru_cache.cpp [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     * Constructor Function
+     * Accepts 2 parameters key and value for all new instaces of class node
+     * created.
+     * @param key : Key
+     * @param val : Value
+     * @returns NOTHING
+     */
+    node(int key, int val) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:49]:
```suggestion
    node(uint64_t key, uint64_t val) {
```

File: data_structures/lru_cache.cpp [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * Constructor Function
+     * Accepts one parameter, the physical size for LRU.
+     * Creates a doubly circular linked list containg {capacity} number of
+     * nodes.
+     * @param capacity the physical size for LRU as needed for particular
+     * instance.
+     * @returns NOTHING
+     */
+    explicit LRUCache(int capacity) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:81]:
```suggestion
    explicit LRUCache(const uint64_t& capacity) {
```

File: data_structures/lru_cache.cpp [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    /**
+     * get Function
+     * Accepts one parameter, the key whose value needs to be returned.
+     * If key requested is not present return -1.
+     * @param key value for this key is to be returned.
+     * @returns NOTHING
+     */
+    int get(int key) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:102]:
```suggestion
    int get(uint64_t key) {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by proRamLOGO** [comment]
Up-to-date

**Comment by proRamLOGO** [comment]
@Panquesito7 please review.

File: data_structures/lru_cache.cpp [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    /**
+     * get Function
+     * Accepts one parameter, the key whose value needs to be returned.
+     * If key requested is not present return -1.
+     * @param key value for this key is to be returned.
+     * @returns NOTHING
+     */
+    int get(uint64_t key) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:102]:
```suggestion
    uint64_t get(uint64_t key) {
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by Panquesito7** [comment]
https://github.com/TheAlgorithms/C-Plus-Plus/pull/1356#discussion_r511094070

**Comment by proRamLOGO** [comment]
> [#1356 (comment)](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1356#discussion_r511094070)

Done

File: data_structures/lru_cache.cpp [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+ * - unordered_map< int , node* > ///< To keep track od addresses of all nodes
+ * involved in linked list. Methods:-
+ * - LRUCache(int): Constructor Function
+ * - VALUE_TYPE get(int): Function to query on the LRU
+ * - void put(int,int): Function to add/update a key in LRU.
+ */
+class LRUCache {
+ public:
+    node *root;
+    std::unordered_map<int, node *> addr;
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:70]:
```suggestion
    std::unordered_map<uint64_t, node *> addr;
```

File: data_structures/lru_cache.cpp [hunk 28]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    /**
+     * get Function
+     * Accepts one parameter, the key whose value needs to be returned.
+     * If key requested is not present return -1.
+     * @param key value for this key is to be returned.
+     * @returns the repective value for requested key. If key is not found then it returns -1.
+     */
+    uint64_t get(uint64_t key) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:102]:
```suggestion
    uint64_t get(const uint64_t& key) {
```

File: data_structures/lru_cache.cpp [hunk 29]
```diff
@@ -0,0 +1,176 @@
...
+     * put Function
+     * Accepts two parameters key and value. Updates the value of the key if the
+     * key exists. Otherwise, adds the key-value pair to the cache. If the
+     * number of keys exceeds the capacity from this operation, the function
+     * evicts the least recently used key.
+     * @param key the key to be added or updated.
+     * @param value value for the respective key to be stored.
+     * @returns NOTHING
+     */
+    void put(uint64_t key, const uint64_t &value) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:137]:
```suggestion
    void put(const uint64_t& key, const uint64_t &value) {
```

File: data_structures/lru_cache.cpp [hunk 30]
```diff
@@ -0,0 +1,176 @@
...
+
+    /**
+     * Constructor Function
+     * Accepts 2 parameters key and value for all new instaces of class node
+     * created.
+     * @param key : Key
+     * @param val : Value
+     * @returns NOTHING
+     */
+    node(uint64_t key, uint64_t val) {
```
Comment (Panquesito7) [inline data_structures/lru_cache.cpp:49]:
```suggestion
    node(const uint64_t& key, const uint64_t& val) {
```

**Review by Panquesito7 — COMMENTED** [review]
👍 Please add self-tests to ensure the algorithm works as excepted.

**Comment by proRamLOGO** [comment]
> 👍 Please add self-tests to ensure the algorithm works as excepted.

Already added.

**Comment by Panquesito7** [comment]
See the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for how to add self-test cases.

**Pull request closed without merging**